cfgov/v1/storage.py

```diff
@@ -0,0 +1,120 @@
+    some filename "test.txt"; the source file was "test.txt" and the file on
+    the storage backend will be "test.txt". If a user uses the Wagtail UI to
+    replace the document's file with a new file "test2.txt", the change in
+    1.13 tries to clean up the old file by deleting "test.txt" from the storage
+    backend after "test2.txt" has been stored.
+
+    The Wagtail logic looks roughly like this:
+
+        - user uploads a new file for an existing document/image
+        - store the new file to the storage backend
+        - delete the document/image's old file from the storage backend
+
+    This logic is sound as long as it is safe to assume that new uploads with
+    identical filenames never overwrite each other and always get assigned a
+    unique filename when stored. However, as described above, this may not
+    always be the case.
+
+    If, instead, files get overwritten, then the final delete that happens
+    will in fact delete the new file that's just been stored.
+
+    This mixin attempts to prevent this from happening. It uses the Python
+    stack inspect capability to identify if a delete call is being invoked
+    from the Wagtail document/image edit views, and, if so, skips that delete.
+    """
+    def delete(self, name):
+        stack = inspect.stack()
+
+        if self._should_skip_wagtail_delete(stack):
+            return
+
+        super(WagtailOverwritingStorageMixin, self).delete(name)
+
+    def _should_skip_wagtail_delete(self, stack):
+        """Given a call stack, determine whether delete should be skipped."""
+
+        # See Python inspect documentation for structure of stack object:
+        # https://docs.python.org/2/library/inspect.html#the-interpreter-stack
+        caller_frame_record = stack[1]
+        caller_frame_object = caller_frame_record[0]
+        caller_frame_function_name = caller_frame_record[3]
+
+        caller_module = inspect.getmodule(caller_frame_object)
+        caller_module_name = caller_module.__name__
+
+        called_from = caller_module_name + '.' + caller_frame_function_name
+
+        # Only skip delete if it is being called by the wagtaildocs.edit or
+        # wagtailimages.edit views. We want deletes called from other places
+        # to work as expected.
+        skip_for_these_callers = (
+            'wagtail.wagtaildocs.views.documents.edit',
+            'wagtail.wagtailimages.views.images.edit',
+        )
+
+        if called_from not in skip_for_these_callers:
+            return
+
+        # We know this delete is being called from by specific Wagtail views of
+        # interest. Grab the variables from the stack that contain the filename
+        # of the document/image being replaced and the filename of the new
+        # document/image with which it is being replaced.
+        original_file = caller_frame_object.f_locals['original_file']
+        doc = caller_frame_object.f_locals['doc']
```

Comment:
It says above this works with images and docs, but it's looking for a `doc` local, which [`wagtail.wagtaildocs.views.documents.delete` has](https://github.com/wagtail/wagtail/blob/stable/1.13.x/wagtail/wagtaildocs/views/documents.py#L182) but [`wagtail.wagtailimages.views.images.delete` does not](https://github.com/wagtail/wagtail/blob/stable/1.13.x/wagtail/wagtailimages/views/images.py#L225) (and images aren't addressed in any of the tests). Is that correct? Does this only work with the document views, or should it work with both? The Wagtail behavior is the same for documents and images.

Reply:
Yes, nice catch! It should work with both. If we come to some consensus about moving forward with this approach I can fix this and add additional tests.